src/integer_mod_q/mat_zq/solve.rs

```diff
@@ -58,44 +62,314 @@ impl MatZq {
+
+        // Saves the permutation of the gaussian elimination.
+        let mut permutation: Vec<i64> = vec![];
+        for i in 0..self.get_num_rows() {
+            permutation.push(i);
+        }
+        for column_nr in 0..self.get_num_columns() {
+            if !indices.is_empty() && indices[indices.len() - 1].0 >= self.get_num_rows() {
+                break;
+            }
+            let used_rows: Vec<i64> = indices.iter().map(|(row, _)| *row).collect();
+
+            // Try to solve the system under the current modulus.
+            if let Some((row_nr, inv)) = find_invertible_entry_column(
+                &matrix_identity_base_gauss,
+                column_nr,
+                used_rows.clone(),
+            ) {
+                let row = inv * matrix_identity_base_gauss.get_row(row_nr).ok()?;
+                matrix_identity_base_gauss.set_row(row_nr, &row, 0).ok()?;
+
+                // Set all other entries in that column to `0` (gaussian elimination).
+                for row_nr in
+                    (0..matrix_identity_base_gauss.get_num_rows()).filter(|x| *x != row_nr)
+                {
+                    let old_row = matrix_identity_base_gauss.get_row(row_nr).ok()?;
+                    let entry: Z = old_row.get_entry(0, column_nr).ok()?;
+                    let new_row = &old_row - entry * &row;
+                    matrix_identity_base_gauss
+                        .set_row(row_nr, &new_row, 0)
+                        .ok()?;
+                }
+
+                if row_count != row_nr {
+                    matrix_identity_base_gauss
+                        .swap_rows(row_count, row_nr)
+                        .unwrap();
+
+                    permutation.swap(row_count.try_into().unwrap(), row_nr.try_into().unwrap());
+                }
+
+                // Save the position of `1` for that row.
+                indices.push((row_count, column_nr));
+                row_count += 1;
+            // Search for a not invertible entry to divide the modulus.
+            } else if let Some(row_nr) =
+                find_not_invertible_entry_column(&matrix_identity_base_gauss, column_nr, used_rows)
+            {
+                // Factorize the modulus with the found entry, compute solutions
+                // for the system under the split modulus and use the
+                // Chinese Remainder Theorem to compute a solution based on these
+                // sub-solutions.
+                let entry: Z = matrix_identity_base_gauss
+                    .get_entry(row_nr, column_nr)
+                    .ok()
+                    .unwrap();
+                self.factorization_and_crt(y, entry);
+            }
         }
 
-        // check if we have computed a solution and return `None`
-        if &(self * &out) != y {
+        // Return `None` if the matrix has no full rank.
+        if indices.is_empty()
+            || indices[indices.len() - 1].0 + 1 < matrix_identity_base_gauss.get_num_rows()
```

Comment:
is this test actually correct? not quite shure, but at least not easy to understand why this checks the desired case

Reply:
We can not have matrices that are empty, so if indices is empty, the matrix has rank 0 and is therefore not full rank.
Otherwise the matrix is full rank if there is a 1 in the last row of the row echolon form of the matrix, which is tested in the second case.